### PR TITLE
Vulnerability Scanner - Update re-scan triggers in cluster

### PR DESCRIPTION
**Pull request by sebasfalcone**
|Related issue|
|---|
| #23624 |

## Description
This PR updates the triggers used to re-scan the agents

### CODE REVIEW

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp [hunk 1]
```diff
@@ -214,95 +216,138 @@ void VulnerabilityScannerFacade::initWazuhDBEventSubscription()
...
+
+    // Check cluster node name changes
+    const auto& clusterNodeName = PolicyManager::instance().getClusterNodeName();
+    if (std::string clusterNodeLastName; stateDB.get(CLUSTER_NODE_NAME_KEY, clusterNodeLastName))
+    {
+        // If the cluster node name changed, perform a scan
+        if (clusterNodeLastName.compare(clusterNodeName) != 0)
+        {
+            logDebug1(WM_VULNSCAN_LOGTAG, "Cluster node name changed (manager re-scan needed)");
+            m_managerAction = ActionWrapper::Action::SCAN;
```
Comment (pereyra-m) [inline src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp:294]: Do we need to check `managerDisabledScan` here?

**Review by pereyra-m — COMMENTED** [review]
Great job!

Some minor comments and questions.

**Review**:

- Code: :yellow_circle:
  - Design: it's a good idea to split the actions for agents and manager, instead of using a single atomic variable. There are many questions about the changes just to make sure everything works as expected.
  - Functionality: Tests were added, manual testing will be performed.
  - Complexity: Some changes like removing `handleManagerScanPolicyChange` and using `handlePolicyChanges` reduce the complexity. But some other ones, like the `ActionWrapper` might be hard to understand for another developers.
  - Style: Removing `enum class ManagerScanStatus` might improve readability.
  - Documentation: a detail with the documented code was mentioned.

- Testing: :white_circle:

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp [hunk 2]
```diff
@@ -214,95 +216,138 @@ void VulnerabilityScannerFacade::initWazuhDBEventSubscription()
...
+        logDebug2(WM_VULNSCAN_LOGTAG, "No policy has changed or no action is needed for the agents.");
+    }
+
+    if (m_managerAction == ActionWrapper::Action::SCAN)
+    {
+        logInfo(WM_VULNSCAN_LOGTAG, "Policy changed. Performing re-scan over the manager.");
+
+        nlohmann::json actionData;
+        actionData["action"] = "scanAgent";
+        actionData["agent_info"]["agent_id"] = "000";
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp:334]: No need, due to the nature of the manager's agent, it should always index

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.hpp [hunk 3]
```diff
@@ -159,6 +148,71 @@ class VulnerabilityScannerFacade final : public Singleton<VulnerabilityScannerFa
...
+         * @return ActionWrapper& Reference to the ActionWrapper.
+         */
+        ActionWrapper& operator=(const Action& a)
+        {
+            if (this->action < a)
+            {
+                this->action = a;
+            }
+            return *this;
+        }
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.hpp:199]: I will simply the code here, and make the logic in method that dispatches the actions

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp [hunk 4]
```diff
@@ -214,95 +216,138 @@ void VulnerabilityScannerFacade::initWazuhDBEventSubscription()
...
     if (std::string moduleLastState; stateDB.get(VD_STATE_KEY, moduleLastState))
     {
         // If the module was disabled and now is enabled, set the flag to re-scan all agents.
-        if (moduleLastState == DISABLED && vulnerabilityDetectionEnabled)
+        if (moduleLastState.compare(DISABLED) == 0 && moduleState.compare(ENABLED) == 0)
         {
-            logDebug1(WM_VULNSCAN_LOGTAG, "Vulnerability scanner module was re-enabled (re-scan needed)");
-            m_shouldRescan.store(true);
+            logDebug1(WM_VULNSCAN_LOGTAG, "Vulnerability scanner module was re-enabled (agents re-scan needed)");
+            m_agentsAction = ActionWrapper::Action::SCAN;
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp:227]: I've added this logic to the handlePolicyChanges method

File: src/wazuh_modules/vulnerability_scanner/qa/test_data_policy/007/Readme.md
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/qa/test_data_policy/007/Readme.md:1]: Done

[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/qa/test_data_policy/007/Readme.md:1]: Done

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp (hunk 1 above)
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp:294]: I've added a new condition

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (MiguelazoDS) [reply src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.hpp:199]: Why do we conditionally assign an action here?

File: src/wazuh_modules/vulnerability_scanner/src/policyManager/policyManager.hpp [hunk 5]
```diff
@@ -685,14 +691,11 @@ class PolicyManager final : public Singleton<PolicyManager>
      * This function retrieves the current status of the manager's scan
      * from the configuration and returns it as a ManagerScanStatus enum value.
      *
-     * @return ManagerScanStatus The status of the manager's scan.
-     *         - If the scan is disabled, returns ManagerScanStatus::DISABLED.
-     *         - If the scan is enabled, returns ManagerScanStatus::ENABLED.
+     * @return true if the manager's scan is disabled, false otherwise.
      */
-    ManagerScanStatus getManagerDisabledScan() const
+    bool getManagerDisabledScan() const
```
Comment (MiguelazoDS) [inline src/wazuh_modules/vulnerability_scanner/src/policyManager/policyManager.hpp:696]: I understand the implementation but the name is not clear to me. What about isManagerScanDisabled()? or simillar

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp [hunk 6]
```diff
@@ -214,95 +216,138 @@ void VulnerabilityScannerFacade::initWazuhDBEventSubscription()
 
 void VulnerabilityScannerFacade::vulnerabilityScanPolicyChange(Utils::RocksDBWrapper& stateDB) const
 {
-    const auto vulnerabilityDetectionEnabled = PolicyManager::instance().isVulnerabilityDetectionEnabled();
-
+    // Check if the vulnerability scanner was enabled/disabled
+    const std::string moduleState = PolicyManager::instance().isVulnerabilityDetectionEnabled() ? ENABLED : DISABLED;
```
Comment (MiguelazoDS) [inline src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp:220]: Aren't we adding overhead here using string, why do not simply use the returned boolean value?
The same review applies to all cases that involve ENABLED, DISABLED variables.

File: src/wazuh_modules/vulnerability_scanner/src/policyManager/policyManager.hpp (hunk 5 above)
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/policyManager/policyManager.hpp:696]: Its following the naming of the other getters (get + name of the policy)

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.hpp:199]: I thought I'd changed the code here
```suggestion
        {
            this->action = a;
            return *this;
        }
```

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp (hunk 6 above)
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp:220]: The put methods for the rocksDB wrapper supports the following types:
- std::string
- rocksdb::Slice

The rocksdb::Slice can be constructed via:
- std::string
- std::string_view
- char*
- SliceParts

**Review by pereyra-m — CHANGES_REQUESTED** [review]
Hi @sebasfalcone 

I don't find some changes:
- Logic to dispatch the actions: https://github.com/wazuh/wazuh/pull/24034#discussion_r1643521726
- Adding logs for tests: https://github.com/wazuh/wazuh/pull/24034#discussion_r1635462852
- New condition: https://github.com/wazuh/wazuh/pull/24034#discussion_r1635480488


Also, there is a manual test that keeps failing: `Disabling/enabling the cluster with the manager scan disabled`. Maybe it i s going to be addressed in another issue?

**Comment by sebasfalcone** [comment]
@pereyra-m About [this comment](https://github.com/wazuh/wazuh/pull/24034#pullrequestreview-2126810762), I've pushed the changes. They were rejected because I made a new commit from the web UI

**Review by MiguelazoDS — APPROVED** [review]

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp
Comment (pereyra-m) [inline src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp:1]: I'd add an explicit check of `managerDisabledScan` configuration value in every place that is related to the manager scan.

With the current changes, if I change the node name with the `managerDisabledScan` as DISABLED, the manager is scanned anyway.

**Review by pereyra-m — CHANGES_REQUESTED** [review]
Good job simplifying the logic and adding more specific logs for the tests!
However, there are still some manual tests that should be analyzed.

Partial review.

**Review by pereyra-m — COMMENTED** [review]
Gj!

This is the rest of the review.

**Testing** :red_circle: 

The development branch was installed, and one agent v4.8.0 was connected.

| Prev. Config. | New Config. | Expected | Obtained | Result |
|--------|--------|--------|--------|--------|
| Cluster disabled, manager scan disabled | Cluster enabled, manager scan disabled | Re-scan agent, don't scan manager | Re-scan agent, don't scan manager | :green_circle:  |
| Cluster enabled, manager scan disabled | Cluster disabled, manager scan disabled  | Re-scan agent, don't scan manager | Re-scan agent, don't scan manager  | :green_circle:  |
| Cluster disabled, manager scan disabled | Cluster enabled, manager scan enabled | Re-scan agent, re-scan manager  | Re-scan agent, re-scan manager | :green_circle:  |
| Cluster enabled, manager scan enabled | Cluster disabled, manager scan enabled |  Re-scan agent, re-scan manager | Re-scan agent, re-scan manager | :green_circle:  | 
| Cluster enabled, manager scan enabled |  Cluster enabled, manager scan enabled, node name changed | Re-scan manager | Re-scan manager | :green_circle: |
| Cluster enabled, manager scan enabled |  Cluster enabled, manager scan disabled, node name changed | No scan | No scan  | :green_circle:  |
| Cluster enabled, manager scan disabled |  Cluster enabled, manager scan disabled, node name changed | No scan | Re-scan manager  | :red_circle:   |
| Cluster enabled, manager scan disabled |  Cluster enabled, manager scan disabled, cluster name changed | Re-scan agent, don't scan manager | Re-scan agent, don't scan manager  | :green_circle:   |

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp [hunk 7]
```diff
@@ -214,95 +216,140 @@ void VulnerabilityScannerFacade::initWazuhDBEventSubscription()
...
+    }
+    stateDB.put(CLUSTER_STATE_KEY, clusterState);
+
+    // Check cluster node name changes
+    const auto& clusterNodeName = PolicyManager::instance().getClusterNodeName();
+    if (std::string clusterNodeLastName; stateDB.get(CLUSTER_NODE_NAME_KEY, clusterNodeLastName))
+    {
+        // If the cluster node name changed, perform a scan
+        // Note: If the manager was disabled (Cleanup needed), we don't need to re-scan the manager.
+        if (clusterNodeLastName.compare(clusterNodeName) != 0 && m_managerAction != ActionWrapper::Action::CLEANUP)
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp:292]: I will add a new state to reflect this

File: src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/vulnerabilityScannerFacade.cpp:1]: I've added a new value to the action that forces no action. This way we can track this case

**Review by pereyra-m — APPROVED** [review]
LGTM!

Great job:
- Manual testing: the last case was fixed
- Code review: the new mechanism prevents unwanted scans of the manager
- Automated tests: the case that generated problems was added to the QA tests

**Pull request merged**